pkg/scheduler/framework/plugins/volumerestrictions/volume_restrictions.go

```diff
@@ -338,15 +340,115 @@ func (pl *VolumeRestrictions) EventsToRegister() []framework.ClusterEventWithHin
+		{Event: framework.ClusterEvent{Resource: framework.Node, ActionType: framework.Add}, QueueingHintFn: pl.isSchedulableAfterNodeAdded},
 		// Pods may fail to schedule because the PVC it uses has not yet been created.
 		// This PVC is required to exist to check its access modes.
-		{Event: framework.ClusterEvent{Resource: framework.PersistentVolumeClaim, ActionType: framework.Add | framework.Update}},
+		{Event: framework.ClusterEvent{Resource: framework.PersistentVolumeClaim, ActionType: framework.Add | framework.Update}, QueueingHintFn: pl.isSchedulableAfterPersistentVolumeClaimChange},
 	}
 }
 
+// isSchedulableAfterPodDeleted is invoked whenever a pod deleted,
+// It checks whether the deleted pod will conflict with volumes of other pods on the same node
+func (pl *VolumeRestrictions) isSchedulableAfterPodDeleted(logger klog.Logger, pod *v1.Pod, oldObj, newObj interface{}) framework.QueueingHint {
+	oldPod, _, err := util.As[*v1.Pod](oldObj, newObj)
+	if err != nil {
+		logger.Error(err, "unexpected objects in isSchedulableAfterPodDeleted", "oldObj", oldObj, "newObj", newObj)
+		return framework.QueueAfterBackoff
+	}
+
+	if pl.verifyReadWriteOncePodPVCConflict(logger, pod) {
+		return framework.QueueAfterBackoff
+	}
+
+	nodeInfo := framework.NewNodeInfo(oldPod)
+	if !satisfyVolumeConflicts(pod, nodeInfo) {
+		return framework.QueueAfterBackoff
+	}
+
+	return framework.QueueSkip
+}
+
+// isSchedulableAfterNodeAdded is invoked whenever a node added, It checks whether
+// that change made a previously unschedulable pod schedulable.
+func (pl *VolumeRestrictions) isSchedulableAfterNodeAdded(logger klog.Logger, pod *v1.Pod, oldObj, newObj interface{}) framework.QueueingHint {
+	_, newNode, err := util.As[*v1.Node](oldObj, newObj)
+	if err != nil {
+		logger.Error(err, "unexpected objects in isSchedulableAfterNodeAdded", "oldObj", oldObj, "newObj", newObj)
+		return framework.QueueAfterBackoff
+	}
+
+	if pl.verifyReadWriteOncePodPVCConflict(logger, pod) {
+		return framework.QueueAfterBackoff
+	}
+
+	logger.V(4).Info("node was added", "pod", klog.KObj(pod), "node", klog.KObj(newNode))
+	return framework.QueueSkip
+}
+
+// isSchedulableAfterPersistentVolumeClaimChange is invoked whenever a PersistentVolumeClaim changed, It checks whether
+// that change made a previously unschedulable pod schedulable.
+func (pl *VolumeRestrictions) isSchedulableAfterPersistentVolumeClaimChange(logger klog.Logger, pod *v1.Pod, oldObj, newObj interface{}) framework.QueueingHint {
+	oldPersistentVolumeClaim, _, err := util.As[*v1.PersistentVolumeClaim](oldObj, newObj)
+	if err != nil {
+		logger.Error(err, "unexpected objects in isSchedulableAfterPersistentVolumeClaimChange", "oldObj", oldObj, "newObj", newObj)
+		return framework.QueueAfterBackoff
+	}
+
+	if oldPersistentVolumeClaim == nil {
+		return framework.QueueImmediately
+	}
+
+	// The new pvc won't affect pod scheduling because it hasn't bound any pod yet, and the AccessModes of the pvc are immutable,
```

Comment:
> the AccessModes of the pvc are immutable,

Is it? I don't see any validation to make AccessModes immutable.

https://github.com/sanposhiho/kubernetes/blob/95c6295636adaeb06d47b02ff4c705b6aa106606/pkg/apis/core/validation/validation.go#L1984

Reply:
@kubernetes/sig-storage-bugs Is it intended to allow people to change `AccessModes`?
I found one issue that seems related, is this issue still valid?
https://github.com/kubernetes/kubernetes/issues/99399#issuecomment-788202224

Reply:
validation makes the pvc spec immutable [here](https://github.com/sanposhiho/kubernetes/blob/95c6295636adaeb06d47b02ff4c705b6aa106606/pkg/apis/core/validation/validation.go#L2252-L2255).

Reply:
Ah, why I saw the validation for PV 😓 OK, then make sense.